src/main/flight/autopilot.c

```diff
@@ -91,14 +155,210 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
     DEBUG_SET(DEBUG_AUTOPILOT_ALTITUDE, 7, lrintf(altitudeF));
 }
 
-bool isBelowLandingAltitude(void)
-{
-    return getAltitudeCm() < 100.0f * autopilotConfig()->landing_altitude_m;
+void resetPositionControlParams(void) {
+    posHold.distanceCm = 0.0f;
+    posHold.previousDistanceCm = 0;
+    posHold.previousDistanceRoll = 0.0f;
+    posHold.previousVelocityRoll = 0.0f;
+    posHold.previousDistancePitch = 0.0f;
+    posHold.previousVelocityPitch = 0.0f;
+    posHold.previousHeading = attitude.values.yaw * 0.1f;
+    posHold.pitchI = 0.0f;
+    posHold.rollI = 0.0f;
 }
 
-const pidCoefficient_t *getAltitudePidCoeffs(void)
+void resetPositionControl(gpsLocation_t initialTargetLocation) {
+    currentTargetLocation = initialTargetLocation;
+    posHold.justStarted = true;
+    resetPositionControlParams();
+}
+
+bool positionControl(float deadband) {
+
+    // exit if we don't have suitable data
+    if (!STATE(GPS_FIX)) {
+        return false; // cannot proceed without a GPS location
+    }
+
+    if (!canUseGPSHeading
+#ifdef USE_MAG
+        && !compassIsHealthy()
+#endif
+        ) {
+        return false;
+        // If compass is healthy, we must have a Mag, and therefore are OK, even with no GPS Heading
+        // If user allows posHold without a Mag, the IMU must be able to get heading from the GPS
+    }
+
+    // collect initial data values - gpsSol.llh = current gps location
+    int32_t bearing = 0; // degrees * 100
+    const float gpsDataIntervalS = getGpsDataIntervalSeconds(); // interval for current GPS data value 0.01s to 1.0s
+    const float gpsDataIntervalHz = 1.0f / gpsDataIntervalS;
+
+    // get distance and bearing from current location (gpsSol.llh) to target location
+    GPS_distance_cm_bearing(&gpsSol.llh, &currentTargetLocation, false, &posHold.distanceCm, &bearing);
+
+    // at the start, if the quad was moving, it will initially show increasing distance from start point
+    // once it has 'stopped' the PIDs will push back towards home, and the distance away will decrease
+    // it looks a lot better if we reset the target point to the point that we 'pull up' at
+    // otherwise there is a big distance to pull back if we start pos hold while carrying some speed
+    if (posHold.justStarted) {
+        if (posHold.distanceCm < posHold.previousDistanceCm) {
+            currentTargetLocation = gpsSol.llh;
+            resetPositionControlParams();
+            posHold.justStarted = false;
```

Comment:
justStarted can become false only here.
Does it mean that if quad just keeps sliding away the actual pos hold will never kick in?
Lets say you fly downwind, and quad will keep sliding away unless you counteract the wind.

Reply:
Good point.
I was not sure about this, but it works.
Whenever position hold starts, the start point is the present location, so the craft can only move away, at first, and the distance away can only ever increase.  
For position hold to be successful, it must arrest the movement away from the start point.  If it fails to ever arrest the movement away, justStarted will never be true, and the sanity check will become active and pos hold has failed.  In any successful position hold, the activity of the PIDs must at some point reverse the outwards velocity, until at some point the quad starts to fly towards the initiation point.  At that justStarted becomes true, and the 'zero movement' point, or close to it, becomes the active target value for the remainder of the pos hold.
This way when you stop the sticks, the quad arrests as fast as possible and then stays at the point it stopped at.

Reply:
I think `juststarted` becomes true as soon as you click the pos hold switch.
And while it's true the pos hold don't try to stop the quad. So then how quad will stop? Are you just waiting for inertia to stop the quad?

Reply:
No, it works now because when position control is activated, it will immediately slow down any existing drift, and there *always* is some.  So the distance away always is increasing at the start.  Unless the corrections fail totally, the distance away will fall.  When we are no longer gaining distance away, we stay at that point.

In this image, the flat spot to left of centre is where I am pushing the sticks to change position (see RCCommand is non-zero).  Then I stop doing that, and the code re-enters position hold.
Note that the distance in cyan in the third panel is increasing.
The PIDs react strongly, and the distance away slows down.
At some point, shortly after, the distance away reverses, since the code tries to push back to the original target location, about a meter or so behind us.
At this point, justStarted changes state, and the target location is reset.  The red cursor is at this point.  That becomes the new target point.
I have checked this behaviour now, countless times; it is very reliable.
<img width="1433" alt="Screen Shot 2024-10-19 at 12 06 58 pm" src="https://github.com/user-attachments/assets/dcd4684e-cda6-43ad-aa3f-a8ed1b4a8476">